Drt bitbucket integration no a

georgiantanase opened the pull request:
- add bitbucket models;
- use rest api to make requests to get pull requests and commits;
-  transform bitbucket pull requests and commits to octokit(github) pull requests and commits, to reuse code;
- updated the workflow for source control, if bitbucket is used

ccriveteanu committed:
- add bitbucket models
- refactor rest client to reuse code
- try authenticate using app password

Commit:
- add models for BitBucket
- add config in appsettings for the base url of the client
- make the pull request request work

Commit:
- get the links from json response
- get all the pull requests with multiple requests

Commit:
- update workflow to take into account the bitbucket enum
- add commit request
- start the log by getting the pull requests and commits

Commit:
- transform bitbucket pullrequest and commit models to jira pullrequest and models
- finish things up

rungureanu-eq commented on DailyReportWeb/DependencyInjection.cs:
```diff
@@ -23,7 +23,8 @@ public static void Init()
 
-            builder.RegisterServicesFromAssembly<JiraService>();            
+            builder.RegisterServicesFromAssembly<JiraService>();
+            builder.RegisterServicesFromAssembly<BitBucketService>();
```
JiraService and BitBucketService are both in the same assembly. Won't this call registers the same services twice?

rungureanu-eq commented on Equilobe.DailyReport.BL/BitBucket/BitBucketApiUrls.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    public class BitBucketApiUrls
+    {
+        public static string PullRequests(string owner, string repository, string page)
```
i think it would be better to have the page as an integer, instead of having to call ToString() when calling the method

rungureanu-eq commented on Equilobe.DailyReport.SL/BitBucketService.cs:
```diff
@@ -0,0 +1,96 @@
+
+        private BitBucketClient GetClient(Credentials credentials, string baseUrl)
+        {
+            return BitBucketClient.CreateWithBasicAuth(baseUrl, credentials.Username, credentials.Password);
```
get the baseurl directly in this method, to avoid calling multiple times GetBaseUrl()

rungureanu-eq commented on Equilobe.DailyReport.SL/BitBucketService.cs:
```diff
@@ -0,0 +1,96 @@
+
+        private List<Commit> ExtractOutOfRangeCommits(List<Commit> commits, DateTime fromDate, DateTime toDate)
+        {
+            commits.RemoveAll(p => p.CreatedAt < fromDate || p.CreatedAt > toDate);
```
you can use a Where clause to directly return the result, if you want

rungureanu-eq commented on Equilobe.DailyReport.SL/BitBucketService.cs:
```diff
@@ -0,0 +1,96 @@
+            return ExtractOutOfRangeCommits(commits, fromDate, toDate);
+        }
+
+        private bool IsAnyCommitOutOfRange(List<Commit> commits, DateTime fromDate)
```
OutOfRange makes me think of indexes in lists. Outdated sounds better or OlderThan or something like this

rungureanu-eq commented on Equilobe.DailyReport.BL/BitBucket/BitBucketLogHelper.cs:
```diff
@@ -0,0 +1,50 @@
+
+            return new LogEntry
+            {
+                Author = commit.Author.User != null ? commit.Author.User.Links.Html.Href : null,
```
i think you can use ?. 
commit.Author.User?.Links?.Html?.Href

rungureanu-eq commented on Equilobe.DailyReport.BL/BitBucket/BitBucketLogHelper.cs:
```diff
@@ -0,0 +1,50 @@
+        {
+            var log = new Log();
+
+            if (pullRequests != null)
```
the code checks for null pullrequests, but you not for null commits.
i think it you can skip this check (as i see, there's no possibility to have null, just an empty list), and also use the object initializer to instantiate the log

rungureanu-eq commented on Equilobe.DailyReport.Models/BitBucket/Markup.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public enum Markup
+    {
+        markdown,
```
Capital first letter?

georgiantanase replied:
this is the exact string that is received from the request

rungureanu-eq commented on Equilobe.DailyReport.Models/BitBucket/State.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public enum State
+    {
+        MERGED,
```
is it necessary to use capital letters for all of them?

georgiantanase replied:
this is the exact string received from the request

rungureanu-eq replied:
the Parse method has an overload, and you can tell it to ignore the casing

rungureanu-eq commented on Equilobe.DailyReport.SL/SourceControlService.cs:
```diff
@@ -18,27 +18,38 @@ public List<string> GetContributors(SourceControlOptions sourceControlOptions)
-                    return GitHubService.GetAllContributors(sourceControlOptions.Credentials, sourceControlOptions.RepoOwner, sourceControlOptions.Repo)
-                                        .Select(qr => qr.Login)
-                                        .ToList();
+                    GetContributorsGithub(sourceControlOptions);
```
return?

rungureanu-eq commented on Equilobe.DailyReport.SL/SourceControlService.cs:
```diff
@@ -18,27 +18,38 @@ public List<string> GetContributors(SourceControlOptions sourceControlOptions)
             return null;
         }
+
+        private List<string> GetContributorsGithub(SourceControlOptions sourceControlOptions)
```
rename to GetGithubContributors

rungureanu-eq commented on Equilobe.DailyReport.SL/SourceControlService.cs:
```diff
@@ -18,27 +18,38 @@ public List<string> GetContributors(SourceControlOptions sourceControlOptions)
+                .ToList();
+        }
+
+        private List<string> GetConstributorsSVN(SourceControlOptions sourceControlOptions)
```
rename to GetSvnContributors

rungureanu-eq commented on JiraReporter/DependencyInjection.cs:
```diff
@@ -23,11 +23,13 @@ public static void Init()
             var builder = new ContainerBuilder();
 
             builder.RegisterServicesFromAssembly<JiraService>();
+            builder.RegisterServicesFromAssembly<BitBucketService>();
```
the same question related to double registration

rungureanu-eq commented on JiraReporter/SourceControl/BitBucketSourceControl.cs:
```diff
@@ -0,0 +1,33 @@
+
+        public override Log CreateLog()
+        {
+            var context = new SourceControlContext { SourceControlOptions = Policy.SourceControlOptions, FromDate = Options.FromDate, ToDate = Options.ToDate };
```
each prorperty on a new line

rungureanu-eq commented on SvnLogReporter/BitBucketReport.cs:
```diff
@@ -0,0 +1,35 @@
+
+        public override Log CreateLog()
+        {
+            var context = new SourceControlContext { SourceControlOptions = Policy.SourceControlOptions, FromDate = Options.FromDate, ToDate = Options.ToDate };
```
each property on a new line

rungureanu-eq commented on JiraReporter/SourceControl/BitBucketSourceControl.cs:
```diff
@@ -0,0 +1,33 @@
+        public override Log CreateLog()
+        {
+            var context = new SourceControlContext { SourceControlOptions = Policy.SourceControlOptions, FromDate = Options.FromDate, ToDate = Options.ToDate };
+            var log = BitBucketService.GetLog(context);
```
return directly

Commit:
- refactor code

rungureanu-eq commented on Equilobe.DailyReport.Models/BitBucket/Summary.cs:
```diff
@@ -0,0 +1,33 @@
+            {
+                try
+                {
+                    return (Markup)Enum.Parse(typeof(Markup), MarkupType);
```
you could create an extension method AsEnum<>() for this parsing.

ccriveteanu committed:
- refactor code

Commit:
- make enum types non-nullable

Sebyd commented on Equilobe.DailyReport.SL/SourceControlService.cs:
```diff
@@ -18,27 +18,38 @@ public List<string> GetContributors(SourceControlOptions sourceControlOptions)
+                    return GetSVNContributors(sourceControlOptions);
+
+                if (sourceControlOptions.Type == SourceControlType.BitBucket)
+                    throw new NotImplementedException();
```
Why is this not implemented?

Commit:
- get contributors for bit bucket

georgiantanase commented on Equilobe.DailyReport.BL/BitBucket/BitBucketApiUrls.cs:
```diff
@@ -0,0 +1,15 @@
+        }
+
+        public static string Commits(string owner, string repository, int page)
+        {
```
It doesn't work, I tried it before. It seems there is actually something written in the api docs about this

Several 2.0 API resources that return collections of objects all support a single, shared, generic querying language that is used to filter down a result set.
Currently, the endpoints which support filtering and sorting include:
/2.0/repositories/{username}
/2.0/repositories/{username}/{slug}/refs
/2.0/repositories/{username}/{slug}/refs/branches
/2.0/repositories/{username}/{slug}/refs/tags
/2.0/repositories/{username}/{slug}/forks
/2.0/repositories/{username}/{slug}/src
/2.0/repositories/{username}/{slug}/issues
/2.0/repositories/{username}/{slug}/pullrequests

So the filtering can't happen to commits.

Commit:
- get author display name instead of the link

Commit:
Merge branch 'DRT-bitbucket-integration-no-a' of https://github.com/equilobe/DailyReportingTools into DRT-bitbucket-integration-no-a

Commit:
- add bitbucket option in the drt web page

Sebyd committed:
- added null check on BitBucketLogHelper
- refactoring GetPullRequests

Sebyd committed:
- fixed typo

rungureanu-eq commented on JiraReporter/SourceControlProcessor.cs:
```diff
@@ -34,10 +35,15 @@ public static Log GetSourceControlLog(JiraReport report)
         public static List<JiraPullRequest> GetPullRequests(Log log)
         {
             var pullRequests = new List<JiraPullRequest>();
```
if (pullReq == null)
return new List();

return pullReq.Select();

Sebyd commented on DailyReportWeb/app/settings.html:
```diff
@@ -266,6 +270,66 @@ <h4>
+                                    Repository Owner:
+                                </label>
+                                <div>
+                                    <input type="text" name="gitRepoOwner" class="form-control" ng-model="data.sourceControlOptions.repoOwner" ng-blur="sourceControlStatus = 'changed'" />
```
bitbucketRepoOwner instead of gitRepoOwner

Sebyd commented on DailyReportWeb/app/settings.html:
```diff
@@ -266,6 +270,66 @@ <h4>
+                                    Repository Name:
+                                </label>
+                                <div>
+                                    <input type="text" name="gitRepoName" class="form-control" ng-model="data.sourceControlOptions.repo" ng-blur="sourceControlStatus = 'changed'" />
```
bitbucketRepoName

Sebyd committed:
- fixed typo in settings.html

Sebyd committed:
- refactoring in SourceControlProcessor.cs

Sebyd committed:
- updated SourceControlService to get contributors from last 2 weeks commits instead of 3 months (this should be updated to a call to bitbucket api for team members)
- removed extra semicolon in SourceControlProcessor.cs

Commit:
- update name of html inputs to bitbucket

Commit:
- create a generic bitbucket page to avoid duplicating code
- update the author from commit to be the display name, in order to match the jira author
- update contributors getting method to an api request which, if null, calls the method for getting the contributors out of the commits
- update links model to contain another links used in contributors request

Commit:
- update jira subtask completed resolution
- update commit revision, having its hash
